### PR TITLE
Add convenience method to send packets only when the other side can receive

### LINKED ISSUE #424: Add hook for only sending packet if the other side can receive
In JamiesWhiteShirt/clothesline-fabric#11, I realized that fabric's network api have no convenience method for mods to opt-out of packet sending to the other side if the other side has no corresponding packet receiver. This causes rejection/client kick from certain servers such as paper. A hook to send only if the packet is identified can ease such conflicts.

### PR DESCRIPTION
**Pull request by liach**
Resolves #424 

Signed-off-by: liach <liach@users.noreply.github.com>

### CODE REVIEW

File: fabric-networking-v0/src/main/java/net/fabricmc/fabric/api/network/ClientSidePacketRegistry.java [hunk 1]
```diff
@@ -78,10 +81,41 @@ default void sendToServer(Packet<?> packet) {
...
 	 * @param buf The packet byte buffer.
 	 */
 	default void sendToServer(Identifier id, PacketByteBuf buf) {
 		sendToServer(id, buf, null);
 	}
+
+	/**
+	 * Send an identifier/buffer-based packet to the server.
+	 *
+	 * <p>This packet won't be send if the server didn't declare it can receive
```
Comment (quat1024) [inline fabric-networking-v0/src/main/java/net/fabricmc/fabric/api/network/ClientSidePacketRegistry.java]: (tiny grammar fix: send -> sent)

**Review by quat1024 — APPROVED** [review]
Looks nice 👍

**Comment by TheGlitch76** [comment]
Maybe consider returning a `boolean` that is `true` if the packet was successfully sent?

**Comment by liach** [comment]
Mods can just call `canServerReceive(Identifier)` before they send the packet instead.

**Comment by JamiesWhiteShirt** [comment]
This approach doesn't "scale" well, so to speak. If we were to introduce another modifier like this, the amount of methods would grow exponentially. I would be content with what we already have with javadocs suggesting to use `canServerReceive/canPlayerReceive` if compatibility is desired.

**Comment by liach** [comment]
err, it's like optional. from your point we just do optional.ispresent and get without methods like orelse because they can be replaced by these two methods. not really: mostly people will forget this check, and note that mods may be developed without the sources of fabric api.

**Comment by JamiesWhiteShirt** [comment]
Optional sends the message that "value present" and "value not present" are two equally valid scenarios, as opposed to a missing value being an exception. I can understand this API design if we are to treat "message supported" and "message not supported" as two equally valid scenarios, and that's probably where we disagree.

I believe that designing for connections where a message may or may not be supported is an exception, and therefore this API doesn't have that much value as a convenience.

I would also like to point out that the use of `canServerReceive/canPlayerReceive` allows avoiding unnecessary computation and serialization to produce a message that will ultimately be discarded, and would probably be the preferred method for me.

**Comment by Player3324** [comment]
In line with the preceding comment I think this is a too much of a niche feature and generally worse than checking canReceive well before assembling the packet.

It may also encourage people calling those new "safer" methods just in case, hiding errors and causing followup issues like loss of synchronization.

**Comment by liach** [comment]
sure, will return boolean to indicate whether sent and accept a packet byte buf supplier/consumer instead.

**Comment by modmuss50** [comment]
I dont think this sloves the core issue you were linking to.

I dont think we need something like this, as like player said its better for the mod to skip over creating the packet, or evening running some logic.

**Comment by liach** [comment]
creation of a packet will be done in the consumer/suppiler the new method takes. sorry, but i just have no time on a pc to update this yet.

**Comment by liach** [comment]
Sure, moved to #426 so that we only emit warnings.

**Pull request closed without merging**